File: IdentityX/testing_framework/project_users_test.go
package testing

import (
	"GoAuth/internal/apierr"
	"crypto/ed25519"
	"encoding/base64"
	"fmt"
	"net/http"
	"testing"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
	"github.com/stretchr/testify/require"
)

func testProjectUsers(t *testing.T, suite *TestSuite) {
	// Create user in parent test context
	client := suite.NewClient(t)
	user := client.WithCredentials("client@mail.com", ValidPassword).
		Register().
		Login().
		CreateProject("test project")

	// Create with the same email as client to prove it's in a different environment
	projectUser := client.WithCredentials("client@mail.com", ValidPassword).
		ProjectRegister(user.projectID).
		ProjectLogin(user.projectID)

	t.Run("ProjectUsersRegister", func(t *testing.T) {
		var wrongFormatID = "wrong-format"
		t.Run("WrongFormatIDProjectRegister", func(t *testing.T) {
			client := suite.NewClient(t)
			client.POST("/projects/" + wrongFormatID + "/register").
				WithBody(map[string]interface{}{
					"email":    user.email,
					"password": user.password,
				}).
				Expect(http.StatusBadRequest).
				HasErrID(apierr.RequestValidationError).
				HasMessage("Validation failed")
		})

		t.Run("InvalidProjectRegister", func(t *testing.T) {
			nonexistentID, err := uuid.NewV7()
			if err != nil {
				t.Fatal(err)
			}
			client := suite.NewClient(t)
			client.POST("/projects/" + nonexistentID.String() + "/register").
				WithBody(map[string]interface{}{
					"email":    user.email,
					"password": user.password,
				}).
				Expect(http.StatusBadRequest).
				HasMessage("invalid reference")
		})

		t.Run("ValidationProjectRegister", func(t *testing.T) {
			for _, spec := range ValidationTests {
				spec := spec
				t.Run(spec.Name, func(t *testing.T) {
					client := suite.NewClient(t)
					client.POST("/projects/" + user.projectID + "/register").
						WithBody(map[string]interface{}{
							"email":    spec.Email,
							"password": spec.Pass,
						}).
						Expect(http.StatusBadRequest).
						HasErrID(apierr.RequestValidationError).
						ValidationError(spec.Errors...)
				})
			}
		})

		t.Run("WeakPasswordValidationProjectRegister", func(t *testing.T) {
			for i, spec := range WeakPasswordTests {
				spec := spec
				i := i
				t.Run(spec.Name, func(t *testing.T) {
					client := suite.NewClient(t)
					client.POST("/projects/" + user.projectID + "/register").
						WithBody(map[string]interface{}{
							"email":    fmt.Sprintf("weak%d@mail.com", i),
							"password": spec.Password,
						}).
						Expect(http.StatusBadRequest).
						HasErrID(apierr.RequestValidationError).
						ValidationError(spec.Errors...)
				})
			}
		})

		t.Run("DuplicateEmailProjectRegister", func(t *testing.T) {
			client := suite.NewClient(t)
			client.POST("/projects/" + user.projectID + "/register").
				WithBody(map[string]interface{}{
					"email":    user.email,
					"password": user.password,
				}).
				Expect(http.StatusConflict).
				HasErrID(apierr.AuthEmailAlreadyUsed).
				HasMessage("error registering user").
				TraceContains("email already in use")
		})

		t.Run("InvalidSchemaTypeRegister", func(t *testing.T) {
			client := suite.NewClient(t)
			client.POST("/projects/"+user.projectID+"/register").
				WithQuery("schema_type", "invalid").
				WithBody(map[string]interface{}{
					"email":    "invalid_schema@email.com",
					"password": user.password,
				}).
				Expect(http.StatusBadRequest).
				HasErrID(apierr.SchemaInvalidSchemaType).
				HasMessage("invalid schema type")
		})

		t.Run("FlowIDSameAsTypeRegister", func(t *testing.T) {
			client := suite.NewClient(t)
			client.POST("/projects/"+user.projectID+"/register").
				WithQuery("schema_type", "context").
				WithQuery("flow_id", "context").
				WithBody(map[string]interface{}{
					"email":    "flow_same_as_type@email.com",
					"password": user.password,
				}).
				Expect(http.StatusBadRequest).
				HasErrID(apierr.SchemaInvalidFlowID).
				HasMessage("flow id can't be the same as a schema type")
		})

		t.Run("MetadataRegisterOnCoreDenied", func(t *testing.T) {
			client := suite.NewClient(t)
			client.POST("/projects/" + user.projectID + "/register").
				WithBody(map[string]interface{}{
					"email":    "metadata_denied@email.com",
					"password": user.password,
					"custom_fields": map[string]interface{}{
						"curso": "Ciência da Computação",
					},
				}).
				Expect(http.StatusBadRequest).
				HasErrID(apierr.SchemaMetadataNotAllowed).
				HasMessage("custom fields are not allowed for core schema")
		})

		t.Run("SuccessProjectRegister", func(t *testing.T) {
			client := suite.NewClient(t)
			client.WithCredentials("new@mail.com", ValidPassword).ProjectRegister(user.projectID)
		})
	})

	t.Run("ProjectUsersLogin", func(t *testing.T) {
		t.Run("WrongPassword", func(t *testing.T) {
			client := suite.NewClient(t)
			client.POST("/projects/" + user.projectID + "/login").
				WithBody(map[string]string{
					"email":    projectUser.email,
					"password": "WrongPass123!",
				}).
				Expect(http.StatusUnauthorized).
				HasErrID(apierr.AuthInvalidCredentials).
				HasMessage("invalid email or password")
		})

		t.Run("WrongEmail", func(t *testing.T) {
			client := suite.NewClient(t)
			client.POST("/projects/" + user.projectID + "/login").
				WithBody(map[string]string{
					"email":    "wrong@mail.com",
					"password": projectUser.password,
				}).
				Expect(http.StatusUnauthorized).
				HasErrID(apierr.AuthInvalidCredentials).
				HasMessage("invalid email or password")
		})

		t.Run("Success", func(t *testing.T) {
			client := suite.NewClient(t)
			client.WithCredentials(projectUser.email, projectUser.password).ProjectLogin(user.projectID)
		})

		t.Run("Logout", func(t *testing.T) {
			client := suite.NewClient(t)
			loggedInUser := client.WithCredentials(projectUser.email, projectUser.password).ProjectLogin(user.projectID)
			loggedInUser.Logout()

			// Try using revoked session
			loggedInUser.POST("/auth/logout").
				Expect(http.StatusUnauthorized).
				HasErrID(apierr.SessionUnauthorized).
				HasMessage("session not found or revoked")
		})
	})

	sessionUser := client.WithCredentials("sessions@mail.com", ValidPassword).
		ProjectRegister(user.projectID).
		ProjectLogin(user.projectID)

	t.Run("ProjectUsersSession", func(t *testing.T) {
		t.Run("ListSessions", func(t *testing.T) {
			// Create a new client with subtest's t for the authenticated client
			// We need to preserve the auth context but use the new t
			authClient := suite.NewClient(t).WithAuth(sessionUser.auth)
			authClient.GET("/sessions").
				Expect(http.StatusOK).
				RequireDataArray().Length().IsEqual(1)
		})

		t.Run("MultipleLoginsSessions", func(t *testing.T) {
			// Create 3 more sessions (we already have 1)
			client2 := suite.NewClient(t)
			client2.WithCredentials(sessionUser.email, sessionUser.password).ProjectLogin(user.projectID)

			client3 := suite.NewClient(t)
			client3.WithCredentials(sessionUser.email, sessionUser.password).ProjectLogin(user.projectID)

			client4 := suite.NewClient(t)
			user4 := client4.WithCredentials(sessionUser.email, sessionUser.password).ProjectLogin(user.projectID)

			arr := user4.GET("/sessions").
				Expect(http.StatusOK).
				RequireDataArray()

			arr.Length().IsEqual(4)

			// Get oldest session ID to revoke
			currentSessionID := arr.Value(0).Object().Value("session_id").String().Raw()
			oldestSessionID := arr.Value(3).Object().Value("session_id").String().Raw()

			// Can't revoke current session
			user4.DELETE("/sessions/" + currentSessionID).
				Expect(http.StatusForbidden).
				HasErrID(apierr.SessionSelfRevokeForbidden).
				HasMessage("cannot revoke the currently active session")

			// Revoke first session
			user4.DELETE("/sessions/" + oldestSessionID).
				Expect(http.StatusOK).
				HasMessage("revoked session")

			// Should have 3 sessions now
			user4.GET("/sessions").
				Expect(http.StatusOK).
				RequireDataArray().Length().IsEqual(3)
		})

		t.Run("RevokeOtherSessions", func(t *testing.T) {
			client := suite.NewClient(t)
			revokeOthers := client.WithCredentials("revoke-others-project@mail.com", ValidPassword).
				ProjectRegister(user.projectID).
				ProjectLogin(user.projectID)

			// Create more sessions
			suite.NewClient(t).WithCredentials(revokeOthers.email, revokeOthers.password).ProjectLogin(user.projectID)
			suite.NewClient(t).WithCredentials(revokeOthers.email, revokeOthers.password).ProjectLogin(user.projectID)

			revokeOthers.DELETE("/sessions/others").
				Expect(http.StatusOK).
				HasMessage("revoked sessions")

			// Should only have current session
			revokeOthers.GET("/sessions").
				Expect(http.StatusOK).
				RequireDataArray().Length().IsEqual(1)
		})

		t.Run("SessionInfo", func(t *testing.T) {
			client := suite.NewClient(t)
			infoUser := client.WithCredentials("session-me@mail.com", ValidPassword).
				ProjectRegister(user.projectID).
				ProjectLogin(user.projectID)

			data := infoUser.GET("/sessions/me").
				Expect(http.StatusOK).
				RequireDataValue()

			spec := map[string]interface{}{
				"refresh_expire_date": AnyNumber{},
				"access": map[string]interface{}{
					"iss": "GoAuth",
					"exp": AnyNumber{},
					"iat": AnyNumber{},
					"jti": AnyUUID{},
					"sub": map[string]interface{}{
						"id":         AnyUUID{},
						"email":      "session-me@mail.com",
						"project_id": AsString{user.projectID, AnyUUID{}},
						"user_type":  "project",
						"metadata":   map[string]interface{}{},
						"session_id": AnyUUID{},
						"user_agent": AnyString{},
						"user_ip":    AnyString{},
					},
				},
			}

			Validate(t, data, spec)
		})

		t.Run("RevokeAllSessions", func(t *testing.T) {
			client := suite.NewClient(t)
			revoked := client.WithCredentials("revoke-all@mail.com", ValidPassword).
				ProjectRegister(user.projectID).
				ProjectLogin(user.projectID)

			// Create more sessions
			suite.NewClient(t).WithCredentials(revoked.email, revoked.password).ProjectLogin(user.projectID)
			suite.NewClient(t).WithCredentials(revoked.email, revoked.password).ProjectLogin(user.projectID)

			revoked.DELETE("/sessions").
				Expect(http.StatusOK).
				HasMessage("revoked sessions")

			// Session should be invalid
			revoked.GET("/sessions").
				Expect(http.StatusUnauthorized).
				HasErrID(apierr.SessionUnauthorized).
				HasMessage("session not found or revoked")
		})
	})

	t.Run("ProjectUserRefresh", func(t *testing.T) {
		client := suite.NewClient(t)
		refreshUser := client.WithCredentials("refresh@mail.com", ValidPassword).
			ProjectRegister(user.projectID).
			ProjectLogin(user.projectID)

		oldAccess := refreshUser.auth.AccessToken
		oldRefresh := refreshUser.auth.RefreshToken

		// Old tokens should be invalid
		oldClient := client.WithAuth(&AuthContext{
			AccessToken:  oldAccess,
			RefreshToken: oldRefresh,
		})

		refreshUser = refreshUser.Refresh()

		require.NotEqual(t, oldClient.auth.AccessToken, refreshUser.auth.AccessToken, "Access token should change after refresh")
		require.NotEqual(t, oldClient.auth.RefreshToken, refreshUser.auth.RefreshToken, "Refresh token should change after refresh")

		oldClient.GET("/sessions").
			Expect(http.StatusUnauthorized).
			HasErrID(apierr.SessionUnauthorized).
			HasMessage("session not found or revoked")
	})

	t.Run("ProjectUsersProjects", func(t *testing.T) {
		client := suite.NewClient(t)
		nested := client.WithCredentials("nested_creator@mail.com", ValidPassword).
			ProjectRegister(user.projectID).
			ProjectLogin(user.projectID)

		t.Run("CreateProject", func(t *testing.T) {
			authClient := suite.NewClient(t).WithAuth(nested.auth)
			authClient.POST("/projects").
				WithBody(map[string]interface{}{
					"project_name": "Test Project",
					"metadata":     map[string]string{"env": "test"},
				}).
				Expect(http.StatusForbidden).
				HasErrID(apierr.AuthNotClient).
				HasMessage("only clients can access this endpoint")
		})

		t.Run("ListProjects", func(t *testing.T) {
			authClient := suite.NewClient(t).WithAuth(nested.auth)
			authClient.GET("/projects").
				Expect(http.StatusForbidden).
				HasErrID(apierr.AuthNotClient).
				HasMessage("only clients can access this endpoint")
		})

		t.Run("GetProject", func(t *testing.T) {
			authClient := suite.NewClient(t).WithAuth(nested.auth)
			authClient.GET("/projects/" + user.projectID).
				Expect(http.StatusForbidden).
				HasErrID(apierr.AuthNotClient).
				HasMessage("only clients can access this endpoint")
		})

		t.Run("UpdateProject", func(t *testing.T) {
			authClient := suite.NewClient(t).WithAuth(nested.auth)
			authClient.PATCH("/projects/" + user.projectID).
				WithBody(map[string]interface{}{
					"project_name": "Updated Project",
					"metadata":     map[string]string{"env": "prod"},
				}).
				Expect(http.StatusForbidden).
				HasErrID(apierr.AuthNotClient).
				HasMessage("only clients can access this endpoint")
		})

		t.Run("GetProjectJWKS", func(t *testing.T) {
			jwksClient := suite.NewClient(t)
			jwksClient.GET("/projects/" + user.projectID + "/.well-known/jwks.json").
				Expect(http.StatusOK).
				JSONObj().Value("keys").
				Array().NotEmpty()
		})

		t.Run("DeleteProject", func(t *testing.T) {
			authClient := suite.NewClient(t).WithAuth(nested.auth)
			authClient.DELETE("/projects/" + user.projectID).
				Expect(http.StatusForbidden).
				HasErrID(apierr.AuthNotClient).
				HasMessage("only clients can access this endpoint")
		})
	})

	t.Run("CryptographicIsolation", func(t *testing.T) {
		client := suite.NewClient(t)
		user := client.WithCredentials("crypto@mail.com", ValidPassword).
			Register().
			Login().
			CreateProject("Crypto Project")

		projectID := user.projectID
		projectUser := client.WithCredentials("user@crypto.com", ValidPassword).
			ProjectRegister(projectID).
			ProjectLogin(projectID)

		accessToken := projectUser.auth.AccessToken

		// 1. Get Project JWKS (Raw)
		jwksResp := client.GET("/projects/" + projectID + "/.well-known/jwks.json").
			Expect(http.StatusOK).
			JSONObj()

		// Extract public key from JWKS
		xBase64 := jwksResp.Value("keys").Array().Value(0).Object().Value("x").String().Raw()
		xBytes, err := base64.RawURLEncoding.DecodeString(xBase64)
		require.NoError(t, err)
		projectPubKey := ed25519.PublicKey(xBytes)

		// 2. Get Global JWKS (Master - Wrapped in Data)
		globalJwksResp := client.GET("/.well-known/jwks.json").
			Expect(http.StatusOK).
			RequireDataObject()

		gxBase64 := globalJwksResp.Value("keys").Array().Value(0).Object().Value("x").String().Raw()
		gxBytes, err := base64.RawURLEncoding.DecodeString(gxBase64)
		require.NoError(t, err)
		masterPubKey := ed25519.PublicKey(gxBytes)

		// 3. Try verifying with Project Pub Key
		_, err = jwt.Parse(accessToken, func(token *jwt.Token) (interface{}, error) {
			return projectPubKey, nil
		})
		require.NoError(t, err, "Token for project %s should be verifiable with its own project JWKS", projectID)

		// 4. Try verifying with Master Pub Key
		_, err = jwt.Parse(accessToken, func(token *jwt.Token) (interface{}, error) {
			return masterPubKey, nil
		})
		require.Error(t, err, "Token MUST NOT be verifiable by Master Pub Key")
	})
}

File: IdentityX/testing_framework/project_users.test.ts
import { beforeAll, describe, expect, test } from "vitest";
import { createClient } from "./helpers/index.js";
import {
    assertErrID,
    assertMessage,
    assertValidationError,
    shouldFail,
} from "./helpers/assert.js";
import { Validate, AnyUUID, AnyDate, AnyNumber, AnyString, AsString } from "./helpers/validate.js";
import { ValidPassword, ValidationTests, WeakPasswordTests } from "./fixtures/auth/testdata.js";
import { importSPKI, jwtVerify } from "jose";

// Error IDs
// RequestValidationError     = fail.ID(0, "REQ", 4, false, ...)      → 0_REQ_0004_D
// AuthEmailAlreadyUsed       = fail.ID(0, "AUTH", 0, false, ...)     → 0_AUTH_0000_D
// AuthInvalidCredentials     = fail.ID(0, "AUTH", 1, false, ...)     → 0_AUTH_0001_D
// AuthNotClient              = fail.ID(0, "AUTH", 2, true, ...)      → 0_AUTH_0002_S
// SessionRevoked             = fail.ID(0, "SESSION", 0, false, ...)  → 0_SESSION_0000_D
// SessionSelfRevokeForbidden = fail.ID(0, "SESSION", 1, true, ...)   → 0_SESSION_0001_S
// TokenReuseIdentified       = fail.ID(0, "TOKEN", 18, false, ...)   → 0_TOKEN_0018_D
// SchemaInvalidSchemaType    = fail.ID(0, "SCHEMA", 2, true, ...)    → 0_SCHEMA_0002_S
// SchemaInvalidFlowID        = fail.ID(0, "SCHEMA", 2, false, ...)   → 0_SCHEMA_0002_D
// SchemaMetadataNotAllowed   = fail.ID(0, "SCHEMA", 7, true, ...)    → 0_SCHEMA_0007_S
const ErrRequestValidation     = "0_REQ_0004_D";
const ErrEmailAlreadyUsed      = "0_AUTH_0000_D";
const ErrInvalidCredentials    = "0_AUTH_0001_D";
const ErrAuthNotClient         = "0_AUTH_0002_S";
const ErrSessionRevoked        = "0_SESSION_0000_D";
const ErrSessionSelfRevoke     = "0_SESSION_0001_S";
const ErrTokenReuseIdentified  = "0_TOKEN_0018_D";
const ErrSchemaInvalidType     = "0_SCHEMA_0002_S";
const ErrSchemaInvalidFlowID   = "0_SCHEMA_0002_D";
const ErrSchemaMetadata        = "0_SCHEMA_0007_S";

// ============================================================================
// Ed25519 helpers
// ============================================================================

function decodeBase64Url(s: string): Buffer {
    const base64 = s.replace(/-/g, "+").replace(/_/g, "/");
    const padded = base64.padEnd(base64.length + (4 - (base64.length % 4)) % 4, "=");
    return Buffer.from(padded, "base64");
}

function verifyEd25519Key(x: string): void {
    const bytes = decodeBase64Url(x);
    expect(bytes.length).toBe(32);
}

// Convert raw Ed25519 public key bytes to a SubjectPublicKeyInfo (SPKI) PEM
// so jose can import it for JWT verification
function rawEd25519ToSPKIPem(rawBytes: Buffer): string {
    // DER prefix for Ed25519 SubjectPublicKeyInfo
    const prefix = Buffer.from("302a300506032b6570032100", "hex");
    const der = Buffer.concat([prefix, rawBytes]);
    const b64 = der.toString("base64");
    return `-----BEGIN PUBLIC KEY-----\n${b64}\n-----END PUBLIC KEY-----`;
}

// ============================================================================
// PROJECT USERS TESTS
// ============================================================================

describe("Project Users", () => {
    let ownerUser: any;
    let projectID: string;
    let projectUserEmail: string;
    let projectUserPassword: string;

    beforeAll(async () => {
        // Create owner and project
        await createClient().withCredentials("client@mail.com", ValidPassword).register();
        ownerUser = await createClient().withCredentials("client@mail.com", ValidPassword).login();
        const project = await ownerUser.post("/projects", {
            project_name: "test project",
            metadata: { env: "test" },
        });
        projectID = project.id;

        // Register project user with same email as client (different environment)
        projectUserEmail = "client@mail.com";
        projectUserPassword = ValidPassword;
        await createClient().http.post(`/projects/${projectID}/register`, {
            email: projectUserEmail,
            password: projectUserPassword,
        });
    });

    // --------------------------------------------------------------------------
    // REGISTER
    // --------------------------------------------------------------------------

    describe("ProjectUsersRegister", () => {
        test("WrongFormatIDProjectRegister", async () => {
            const body = await shouldFail(
                createClient().http.post("/projects/wrong-format/register", {
                    email: "client@mail.com",
                    password: ValidPassword,
                }),
                400
            );
            assertErrID(body, ErrRequestValidation);
            assertMessage(body, "Validation failed");
        });

        test("InvalidProjectRegister", async () => {
            const fakeID = "00000000-0000-7000-8000-000000000000";
            const body = await shouldFail(
                createClient().http.post(`/projects/${fakeID}/register`, {
                    email: "client@mail.com",
                    password: ValidPassword,
                }),
                400
            );
            assertMessage(body, "can't register on a non existant project");
        });

        describe("ValidationProjectRegister", () => {
            for (const spec of ValidationTests) {
                test(spec.name, async () => {
                    const body = await shouldFail(
                        createClient().http.post(`/projects/${projectID}/register`, {
                            email: spec.email,
                            password: spec.pass,
                        }),
                        400
                    );
                    assertErrID(body, ErrRequestValidation);
                    assertValidationError(body, ...spec.errors);
                });
            }
        });

        describe("WeakPasswordValidationProjectRegister", () => {
            for (const [i, spec] of WeakPasswordTests.entries()) {
                test(spec.name, async () => {
                    const body = await shouldFail(
                        createClient().http.post(`/projects/${projectID}/register`, {
                            email: `weak${i}@mail.com`,
                            password: spec.password,
                        }),
                        400
                    );
                    assertErrID(body, ErrRequestValidation);
                    assertValidationError(body, ...spec.errors);
                });
            }
        });

        test("DuplicateEmailProjectRegister", async () => {
            const body = await shouldFail(
                createClient().http.post(`/projects/${projectID}/register`, {
                    email: "client@mail.com",
                    password: ValidPassword,
                }),
                409
            );
            assertErrID(body, ErrEmailAlreadyUsed);
            assertMessage(body, "email already in use");
        });

        test("InvalidSchemaTypeRegister", async () => {
            const body = await shouldFail(
                createClient().http.post(
                    `/projects/${projectID}/register?schema_type=invalid`,
                    { email: "invalid_schema@email.com", password: ValidPassword }
                ),
                400
            );
            assertErrID(body, ErrSchemaInvalidType);
            assertMessage(body, "invalid schema type");
        });

        test("FlowIDSameAsTypeRegister", async () => {
            const body = await shouldFail(
                createClient().http.post(
                    `/projects/${projectID}/register?schema_type=context&flow_id=context`,
                    { email: "flow_same_as_type@email.com", password: ValidPassword }
                ),
                400
            );
            assertErrID(body, ErrSchemaInvalidFlowID);
            assertMessage(body, "flow id can't be the same as a schema type");
        });

        test("MetadataRegisterOnCoreDenied", async () => {
            const body = await shouldFail(
                createClient().http.post(`/projects/${projectID}/register`, {
                    email: "metadata_denied@email.com",
                    password: ValidPassword,
                    custom_fields: { curso: "Ciência da Computação" },
                }),
                400
            );
            assertErrID(body, ErrSchemaMetadata);
            assertMessage(body, "custom fields are not allowed for core schema");
        });

        test("SuccessProjectRegister", async () => {
            await createClient().http.post(`/projects/${projectID}/register`, {
                email: "new@mail.com",
                password: ValidPassword,
            });
        });
    });

    // --------------------------------------------------------------------------
    // LOGIN
    // --------------------------------------------------------------------------

    describe("ProjectUsersLogin", () => {
        test("WrongPassword", async () => {
            const body = await shouldFail(
                createClient().http.post(`/projects/${projectID}/login`, {
                    email: projectUserEmail,
                    password: "WrongPass123!",
                }),
                401
            );
            assertErrID(body, ErrInvalidCredentials);
            assertMessage(body, "invalid email or password");
        });

        test("WrongEmail", async () => {
            const body = await shouldFail(
                createClient().http.post(`/projects/${projectID}/login`, {
                    email: "wrong@mail.com",
                    password: projectUserPassword,
                }),
                401
            );
            assertErrID(body, ErrInvalidCredentials);
            assertMessage(body, "invalid email or password");
        });

        test("Success", async () => {
            await createClient()
                .withCredentials(projectUserEmail, projectUserPassword)
                .projectLogin(projectID);
        });

        test("Logout", async () => {
            const loggedIn = await createClient()
                .withCredentials(projectUserEmail, projectUserPassword)
                .projectLogin(projectID);

            await loggedIn.logout();

            const body = await shouldFail(loggedIn.logout(), 401);
            assertErrID(body, ErrSessionRevoked);
            assertMessage(body, "session not found or revoked");
        });
    });

    // --------------------------------------------------------------------------
    // SESSIONS
    // --------------------------------------------------------------------------

    describe("ProjectUsersSession", () => {
        let sessionUser: any;
        const sessionEmail = "sessions@mail.com";

        beforeAll(async () => {
            await createClient().http.post(`/projects/${projectID}/register`, {
                email: sessionEmail,
                password: ValidPassword,
            });
            sessionUser = await createClient()
                .withCredentials(sessionEmail, ValidPassword)
                .projectLogin(projectID);
        });

        test("ListSessions", async () => {
            const sessions = await sessionUser.get("/sessions");
            expect(sessions).toHaveLength(1);
        });

        test("MultipleLoginsSessions", async () => {
            await createClient().withCredentials(sessionEmail, ValidPassword).projectLogin(projectID);
            await createClient().withCredentials(sessionEmail, ValidPassword).projectLogin(projectID);
            const user4 = await createClient().withCredentials(sessionEmail, ValidPassword).projectLogin(projectID);

            const sessions: any[] = await user4.get("/sessions");
            expect(sessions).toHaveLength(4);

            const currentSessionID = sessions[0].session_id;
            const oldestSessionID  = sessions[3].session_id;

            const forbiddenBody = await shouldFail(user4.del(`/sessions/${currentSessionID}`), 403);
            assertErrID(forbiddenBody, ErrSessionSelfRevoke);
            assertMessage(forbiddenBody, "cannot revoke the currently active session");

            await user4.del(`/sessions/${oldestSessionID}`);

            const updated: any[] = await user4.get("/sessions");
            expect(updated).toHaveLength(3);
        });

        test("RevokeOtherSessions", async () => {
            const email = "revoke-others-project@mail.com";
            await createClient().http.post(`/projects/${projectID}/register`, {
                email,
                password: ValidPassword,
            });
            const revokeOthers = await createClient()
                .withCredentials(email, ValidPassword)
                .projectLogin(projectID);

            await createClient().withCredentials(email, ValidPassword).projectLogin(projectID);
            await createClient().withCredentials(email, ValidPassword).projectLogin(projectID);

            await revokeOthers.del("/sessions/others");

            const sessions: any[] = await revokeOthers.get("/sessions");
            expect(sessions).toHaveLength(1);
        });

        test("SessionInfo", async () => {
            const email = "session-me@mail.com";
            await createClient().http.post(`/projects/${projectID}/register`, {
                email,
                password: ValidPassword,
            });
            const infoUser = await createClient()
                .withCredentials(email, ValidPassword)
                .projectLogin(projectID);

            const data = await infoUser.get("/sessions/me");

            Validate(data, {
                refresh_expire_date: AnyNumber,
                access: {
                    iss: AsString(projectID, AnyUUID),
                    exp: AnyNumber,
                    iat: AnyNumber,
                    jti: AnyUUID,
                    sub: {
                        id:         AnyUUID,
                        email:      email,
                        project_id: AsString(projectID, AnyUUID),
                        user_type:  "project",
                        metadata:   {},
                        session_id: AnyUUID,
                        user_agent: AnyString,
                        user_ip:    AnyString,
                    },
                },
            });
        });

        test("RevokeAllSessions", async () => {
            const email = "revoke-all@mail.com";
            await createClient().http.post(`/projects/${projectID}/register`, {
                email,
                password: ValidPassword,
            });
            const revoked = await createClient()
                .withCredentials(email, ValidPassword)
                .projectLogin(projectID);

            await createClient().withCredentials(email, ValidPassword).projectLogin(projectID);
            await createClient().withCredentials(email, ValidPassword).projectLogin(projectID);

            await revoked.del("/sessions");

            const body = await shouldFail(revoked.get("/sessions"), 401);
            assertErrID(body, ErrSessionRevoked);
            assertMessage(body, "session not found or revoked");
        });
    });

    // --------------------------------------------------------------------------
    // REFRESH REUSE
    // --------------------------------------------------------------------------

    test("ProjectUserRefreshReuse", async () => {
        const email = "refresh@mail.com";
        await createClient().http.post(`/projects/${projectID}/register`, {
            email,
            password: ValidPassword,
        });
        const refreshUser = await createClient()
            .withCredentials(email, ValidPassword)
            .projectLogin(projectID);

        const oldAuth = refreshUser.auth!;
        const refreshed = await refreshUser.refresh();

        expect(refreshed.auth!.accessToken).not.toBe(oldAuth.accessToken);
        expect(refreshed.auth!.refreshToken).not.toBe(oldAuth.refreshToken);

        // Old tokens should be rejected
        const oldClient = createClient().withAuth(oldAuth);
        const body = await shouldFail(oldClient.get("/sessions"), 401);
        assertErrID(body, ErrTokenReuseIdentified);
        assertMessage(body, "refresh token reuse not allowed");
    });

    // --------------------------------------------------------------------------
    // PROJECT ACCESS DENIED FOR PROJECT USERS
    // --------------------------------------------------------------------------

    describe("ProjectUsersProjects", () => {
        let nestedUser: any;

        beforeAll(async () => {
            const email = "nested_creator@mail.com";
            await createClient().http.post(`/projects/${projectID}/register`, {
                email,
                password: ValidPassword,
            });
            nestedUser = await createClient()
                .withCredentials(email, ValidPassword)
                .projectLogin(projectID);
        });

        const expectForbidden = async (promise: Promise<any>) => {
            const body = await shouldFail(promise, 403);
            assertErrID(body, ErrAuthNotClient);
            assertMessage(body, "only clients can access this endpoint");
        };

        test("CreateProject", async () => {
            await expectForbidden(
                nestedUser.post("/projects", { project_name: "Test Project", metadata: { env: "test" } })
            );
        });

        test("ListProjects", async () => {
            await expectForbidden(nestedUser.get("/projects"));
        });

        test("GetProject", async () => {
            await expectForbidden(nestedUser.get(`/projects/${projectID}`));
        });

        test("UpdateProject", async () => {
            await expectForbidden(
                nestedUser.patch(`/projects/${projectID}`, {
                    project_name: "Updated Project",
                    metadata: { env: "prod" },
                })
            );
        });

        test("GetProjectJWKS", async () => {
            const body = await shouldFail(
                nestedUser.get(`/projects/${projectID}/.well-known/jwks.json`),
                403
            );
            assertErrID(body, ErrAuthNotClient);
        });

        test("DeleteProject", async () => {
            await expectForbidden(nestedUser.del(`/projects/${projectID}`));
        });
    });

    // --------------------------------------------------------------------------
    // CRYPTOGRAPHIC ISOLATION
    // --------------------------------------------------------------------------

    test("CryptographicIsolation", async () => {
        // Create a fresh owner and project for isolation test
        await createClient().withCredentials("crypto@mail.com", ValidPassword).register();
        const cryptoOwner = await createClient().withCredentials("crypto@mail.com", ValidPassword).login();
        const cryptoProject = await cryptoOwner.post("/projects", {
            project_name: "Crypto Project",
            metadata: { env: "test" },
        });
        const cryptoProjectID = cryptoProject.id;

        // Register and login a project user
        await createClient().http.post(`/projects/${cryptoProjectID}/register`, {
            email: "user@crypto.com",
            password: ValidPassword,
        });
        const cryptoUser = await createClient()
            .withCredentials("user@crypto.com", ValidPassword)
            .projectLogin(cryptoProjectID);

        const accessToken = cryptoUser.auth!.accessToken;

        // 1. Get Project JWKS
        const projectJwksRes = await cryptoOwner.http.get(
            `/projects/${cryptoProjectID}/.well-known/jwks.json`,
            { headers: { Cookie: `access_token=${cryptoOwner.auth!.accessToken}; refresh_token=${cryptoOwner.auth!.refreshToken}` } }
        );
        const projectXBase64: string = projectJwksRes.data.keys[0].x;
        const projectKeyBytes = decodeBase64Url(projectXBase64);
        verifyEd25519Key(projectXBase64);

        // 2. Get Global JWKS
        const globalJwksRes = await createClient().http.get("/.well-known/jwks.json");
        const globalXBase64: string = globalJwksRes.data.keys[0].x;
        const globalKeyBytes = decodeBase64Url(globalXBase64);
        verifyEd25519Key(globalXBase64);

        // 3. Token should verify with project key
        const projectPem = rawEd25519ToSPKIPem(projectKeyBytes);
        const projectKey = await importSPKI(projectPem, "EdDSA");
        await expect(
            jwtVerify(accessToken, projectKey)
        ).resolves.toBeDefined();

        // 4. Token must NOT verify with master key
        const masterPem = rawEd25519ToSPKIPem(globalKeyBytes);
        const masterKey = await importSPKI(masterPem, "EdDSA");
        await expect(
            jwtVerify(accessToken, masterKey)
        ).rejects.toThrow();
    });
});